c2rust-refactor/scripts/upgrade_ptr_to_ref.lua

```diff
@@ -0,0 +1,708 @@
+    local stmt_kind = stmt:get_kind()
+    local cfg = self.node_id_cfgs[stmt:get_id()]
+
+    if cfg then
+        if cfg:is_del() then
+            return {}
+        elseif cfg:is_byteswap() and stmt:get_kind() == "Semi" then
+            local expr = stmt:get_node()
+            local lhs_id = cfg.extra_data[1]
+            local rhs_id = cfg.extra_data[2]
+            local lhs = expr:find_subexpr(lhs_id)
+            local rhs = expr:find_subexpr(rhs_id)
+
+            if lhs and rhs then
+                rhs:to_method_call("swap_bytes", {rhs})
+
+                local assign_expr = self.tctx:assign_expr(lhs, rhs)
+
+                stmt:to_semi(assign_expr)
+            end
+        end
+    end
+
+    walk(stmt)
+
+    return {stmt}
+end
+
+function Visitor:visit_struct_field(field)
+    local field_id = field:get_id()
+    local field_ty = field:get_ty()
+    local conversion_cfg = self.node_id_cfgs[field_id]
+
+    if conversion_cfg then
+        local field_ty_kind = field_ty:get_kind()
+
+        -- *mut T -> Box<T>, or Box<[T]> or Option<Box<T>> or Option<Box<[T]>>
+        if field_ty_kind == "Ptr" then
+            field:set_ty(upgrade_ptr(field_ty, conversion_cfg))
+        -- [*mut T; X] -> [Box<T>; X] or [Box<[T]>; X] or [Option<Box<T>>; X]
+        -- or [Option<Box<[T]>; X]
+        elseif field_ty_kind == "Array" then
+            local inner_ty = field_ty:get_tys()[1]
+
+            if inner_ty:get_kind() == "Ptr" then
+                inner_ty = upgrade_ptr(inner_ty, conversion_cfg)
+
+                field_ty:set_tys{inner_ty}
+                field:set_ty(field_ty)
+            end
+        end
+    end
+end
+
+function is_null_ptr(expr)
+    if expr and expr:get_kind() == "Cast" then
+        local cast_expr = expr:get_exprs()[1]
+        local cast_ty = expr:get_ty()
+        local lit = cast_expr:get_lit()
+
+        if lit and lit:get_value() == 0 and cast_ty:get_kind() == "Ptr" then
+            return true
+        end
+    end
+
+    return false
+end
+
+function Visitor:visit_local(locl)
+    local local_id = locl:get_id()
+    local conversion_cfg = self.node_id_cfgs[local_id]
+
+    -- let x: *mut T = 0 as *mut T; -> let mut x = None;
+    -- or let mut x;
+    if conversion_cfg then
+        if conversion_cfg:is_opt_any() then
+            local init = locl:get_init()
+
+            if is_null_ptr(init) then
+                init:to_ident_path("None")
+
+                locl:set_ty(nil)
+                locl:set_init(init)
+            end
+        elseif conversion_cfg:is_box_any() then
+            local init = locl:get_init()
+
+            if is_null_ptr(init) then
+                locl:set_ty(nil)
+                locl:set_init(nil)
+            end
+        end
+
+        local pat_hirid = self.tctx:nodeid_to_hirid(locl:get_pat_id())
+
+        self:add_var(pat_hirid, Variable.new(local_id, true))
+    end
+end
+
+refactor:transform(
+    function(transform_ctx)
+        local node_id_cfgs = {
```

Comment:
Let's at least put a TODO here that these ids should be passed to this transform, and they are specific to the test case.